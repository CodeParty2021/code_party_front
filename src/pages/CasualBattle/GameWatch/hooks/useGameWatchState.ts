import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import {
  ResultType,
  useFetchResult,
} from "hooks/ResultAPIHooks/useFetchResult";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";

export type IResponse = {
  isAnalyzing: boolean;
  analyzingError: boolean;
  result?: ResultType;
  json?: string;
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: result, fetchResult } = useFetchResult();
  const { data: json, fetchJson } = useFetchJson();
  const navigate = useNavigate();

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/casual-battle");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchResult(room.info?.analyzingResult.resultId);
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  const _exitBtnHandler = () => {
    navigate("/casual-battle/waiting-room");
  };

  return {
    isAnalyzing: !room.info?.analyzingResult?.resultId,
    analyzingError: Boolean(room.info?.analyzingResult?.error),
    result: result,
    json: json,
    exitBtnHandler: _exitBtnHandler,
  };
};
