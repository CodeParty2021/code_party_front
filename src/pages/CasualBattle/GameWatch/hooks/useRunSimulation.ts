import { useEffect } from "react";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useRunCodes } from "hooks/CodeAPIHooks/useRunCodes";

export type IResponse = {};

export const useRunSimulation = (): IResponse => {
  const { room, isHost, updateRoomInfo } = useRoomSync();
  const {
    data: simulationResult,
    error: simulationError,
    runCodes,
  } = useRunCodes();

  //初期処理
  useEffect(() => {
    //シミュレーションの実行(ホストのみ実行)
    if (isHost) {
      const codes: string[] = [];
      room.sortedKeysOfMembers.forEach((key) => {
        const member = room.members[key];
        if (member.codeId) codes.push(member.codeId);
      });
      runCodes(...codes);
    }
  }, []);

  //シミュレーション結果を受信した場合
  useEffect(() => {
    if (simulationResult) {
      updateRoomInfo({
        analyzingResult: {
          resultId: simulationResult.jsonId,
        },
      });
    }
  }, [simulationResult]);

  //シミュレーション中にエラーが発生した場合
  useEffect(() => {
    if (simulationError) {
      updateRoomInfo({
        analyzingResult: {
          error: "シミュレーション実行中にエラーが発生しました。",
        },
      });
    }
  }, [simulationError]);

  return {};
};
