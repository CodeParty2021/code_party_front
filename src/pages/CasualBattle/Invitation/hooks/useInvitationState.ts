import { useEffect, useState } from "react";
import { Params, useNavigate, useParams } from "react-router-dom";
import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";

export type IResponse = {
  isLogin: boolean;
  errorMessage?: string;
  roomId?: string;
  enterBtnDisabled: boolean;
  enterBtnClickHandler: () => void;
};

export const useInvitationState = (): IResponse => {
  const { roomId } = useParams<Params<string>>();
  const { isLogin } = useSelector((state: RootState) => state.user);
  const [enterBtnDisabled, setEnterBtnDisabled] = useState(false);
  const { room, enterRoom } = useRoomSync();
  const [errorMessage, setErrorMessage] = useState<string | undefined>(
    undefined
  );
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/casual-battle/waiting-room");
    }
  }, [room.isEntered]);

  //入場ボタン
  const _enterBtnHandler = () => {
    if (typeof roomId === "string") {
      setEnterBtnDisabled(true);
      enterRoom(roomId).catch(() => {
        setEnterBtnDisabled(false);
      });
    } else {
      setErrorMessage("招待リンクが無効です。");
    }
  };

  return {
    isLogin: isLogin,
    errorMessage: errorMessage,
    roomId: roomId,
    enterBtnDisabled: enterBtnDisabled,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
