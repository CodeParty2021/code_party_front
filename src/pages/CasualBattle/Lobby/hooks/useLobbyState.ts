import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  roomCreateBtnDisabled: boolean;
  roomCreateBtnHandler: () => void;
  roomSearchBtnHandler: () => void;
};

export const useLobbyState = (): IResponse => {
  const [roomCreateBtnDisabled, setRoomCreateBtnDisabled] = useState(false);
  const { room, createRoom } = useRoomSync();
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/casual-battle/waiting-room");
    }
  }, [room.isEntered]);

  const _roomCreateBtnHandler = () => {
    setRoomCreateBtnDisabled(true);
    createRoom().catch(() => {
      setRoomCreateBtnDisabled(false);
    });
  };

  const _roomSearchBtnHandler = () => {
    navigate("/casual-battle/search-room");
  };

  return {
    roomCreateBtnDisabled: roomCreateBtnDisabled,
    roomCreateBtnHandler: _roomCreateBtnHandler,
    roomSearchBtnHandler: _roomSearchBtnHandler,
  };
};
