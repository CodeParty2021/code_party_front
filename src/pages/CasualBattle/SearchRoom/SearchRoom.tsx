import React from "react";
import { Link } from "react-router-dom";

type Props = {};

export const CasualBattleSearchRoom: React.FC<Props> = () => {
  return (
    <div>
      <div>ルームIDを入力する</div>
      <input></input>
      <div>
        <Link to="/casual-battle/waiting-room">ルームに入る</Link>
      </div>
    </div>
  );
};
