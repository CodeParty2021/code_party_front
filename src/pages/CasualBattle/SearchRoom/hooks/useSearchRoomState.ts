import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  roomIdTextBoxValue: string;
  roomIdTextBoxChangeHandler: any; // (value: string) => void;にしたいけど，no-unused-varsにひっかかる．
  enterBtnClickHandler: () => void;
};

export const useSearchRoomState = (): IResponse => {
  const { room, enterRoom } = useRoomSync();
  const [roomId, setRoomId] = useState("");
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/casual-battle/waiting-room");
    }
  }, [room.isEntered]);

  const _roomIdTextBoxChangeHandler = (value: string) => {
    setRoomId(value);
  };

  const _enterBtnHandler = () => {
    enterRoom(roomId);
  };

  return {
    roomIdTextBoxValue: roomId,
    roomIdTextBoxChangeHandler: _roomIdTextBoxChangeHandler,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
