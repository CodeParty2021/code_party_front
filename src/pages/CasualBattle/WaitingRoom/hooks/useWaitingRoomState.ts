import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { CodeType, useFetchCodes } from "hooks/CodeAPIHooks/useFetchCodes";
import { UserState, UserAction } from "services/RoomSync/RoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";
import { User } from "services/user/user";

export type IResponse = {
  user: User | null;
  roomInfo: {
    roomId: string;
    host: UserState;
    memberKeys: string[];
    members: { [id: string]: UserState };
    actionKeys: string[];
    actions: { [id: string]: UserAction };
  };
  isHost: boolean;
  status?: "waiting" | "watching";
  ready: boolean;

  readyBtnHandler: () => void;
  readyBtnDisabled: boolean;
  exitBtnHandler: () => void;
  startBtnDisabled: boolean;
  startBtnHandler: () => void;
  kickUserHandler: (userId: string) => void;
  code: {
    codes: CodeType[];
    loading: boolean;
  };
  selectedCodeId: string;
  onChangeSelectedCodeId: (codeId: string) => void;
};

export const useWaitingRoomState = (): IResponse => {
  const {
    room,
    isHost,
    updateRoomInfo,
    updateMember,
    updateOtherMember,
    exitRoom,
  } = useRoomSync();
  const { data: codes = [], loading } = useFetchCodes();
  const [preRoomStatus, setPreRoomStatus] = useState<"waiting" | "watching">(
    "watching"
  );
  const [selectedCodeId, setSelectedCodeId] = useState("");
  const [ready, setReady] = useState(false);
  const [readyBtnDisabled, setReadyBtnDisabled] = useState(true);
  const [startBtnDisabled, setStartBtnDisabled] = useState(true);
  let { user } = useSelector((state: RootState) => state.user);

  const navigate = useNavigate();
  const dummyUser: UserState = {
    displayName: "",
    status: "disconnect",
    ready: false,
  };

  // ユーザ状態を更新
  useEffect(() => {
    updateMember({
      status: "waiting",
    });
  }, []);

  // DB上で退出処理が完了したら退出
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/casual-battle");
    }
  }, [room.isEntered]);

  // 準備ボタンの有効化
  useEffect(() => {
    //コードが選択されたら準備完了ボタンを押せるようにする
    if (selectedCodeId != "") {
      setReadyBtnDisabled(false);
    }
  }, [selectedCodeId]);

  // ルームステータスがwaitingからwatchingになったら観戦画面に遷移
  useEffect(() => {
    if (room.info) {
      if (preRoomStatus == "waiting" && room.info.status == "watching") {
        navigate("/casual-battle/result");
      }
      setPreRoomStatus(room.info?.status);
    }
  }, [room.info?.status]);

  // ready状態をDBに反映
  useEffect(() => {
    updateMember({
      ready: ready,
      codeId: selectedCodeId,
    });
  }, [ready]);

  // ホストの処理
  useEffect(() => {
    if (isHost) {
      if (room.info && room.info.status == "waiting") {
        // 全員が準備完了になったら開始ボタンを有効にする
        let newDisabled = false;
        for (let key of Object.keys(room.members)) {
          if (!room.members[key].ready) {
            newDisabled = true;
            break;
          }
        }
        setStartBtnDisabled(newDisabled);
      }
      // 全員がWaitingRoomに返ってきたらroomの状態をwaitingに更新する
      let newRoomStatus: "waiting" | "watching" = "waiting";
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "watching") {
          newRoomStatus = "watching";
        }
      }
      if (newRoomStatus == "waiting" && room.info?.status == "watching") {
        updateRoomInfo({
          status: "waiting",
          analyzingResult: null,
        });
      }
    }
  }, [room.members, isHost]);

  // kick処理 member.status = kickingの場合そのユーザをキックする
  useEffect(() => {
    if (user) {
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "kicking" && user.id == key) {
          exitRoom();
        }
      }
    }
  }, [room.members]);

  // ---- イベントハンドラ ---- //

  const _readyBtnHandler = () => {
    setReady(!ready);
  };

  const _exitBtnHandler = () => {
    exitRoom();
  };

  const _startBtnHandler = () => {
    updateRoomInfo({
      status: "watching",
    });
  };

  const _onChangeSelectedCodeId = (codeId: string) => {
    setSelectedCodeId(codeId);
  };

  const _kickUserBtnHandler = (userId: string) => {
    updateOtherMember(userId, { status: "kicking" });
  };

  return {
    user: user,
    roomInfo: {
      roomId: room.id ? room.id : "",
      host:
        room.info && room.info.host in room.members
          ? room.members[room.info.host]
          : dummyUser,
      memberKeys: room.sortedKeysOfMembers,
      members: room.members,
      actionKeys: room.sortedKeysOfActions,
      actions: room.actions,
    },
    isHost: isHost,
    status: room.info?.status,
    ready: ready,

    readyBtnHandler: _readyBtnHandler,
    readyBtnDisabled: readyBtnDisabled,
    exitBtnHandler: _exitBtnHandler,
    startBtnDisabled: startBtnDisabled,
    startBtnHandler: _startBtnHandler,
    kickUserHandler: _kickUserBtnHandler,
    code: {
      codes: codes,
      loading: loading,
    },
    selectedCodeId: selectedCodeId,
    onChangeSelectedCodeId: _onChangeSelectedCodeId,
  };
};
