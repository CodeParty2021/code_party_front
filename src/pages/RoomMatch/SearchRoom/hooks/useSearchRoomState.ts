import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";

export type IResponse = {
  roomIdTextBoxValue: string;
  roomIdTextBoxChangeHandler: (value: string) => void;
  enterBtnDisabled: boolean;
  enterBtnClickHandler: () => void;
};

export const useSearchRoomState = (): IResponse => {
  const { room, enterRoom } = useRoomSync();
  const [roomId, setRoomId] = useState("");
  const [enterBtnDisabled, setEnterBtnDisabled] = useState(false);
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  const _roomIdTextBoxChangeHandler = (value: string) => {
    setRoomId(value);
  };

  const _enterBtnHandler = () => {
    setEnterBtnDisabled(true);
    enterRoom(roomId).catch(() => {
      setEnterBtnDisabled(false);
    });
  };

  return {
    roomIdTextBoxValue: roomId,
    roomIdTextBoxChangeHandler: _roomIdTextBoxChangeHandler,
    enterBtnDisabled: enterBtnDisabled,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
