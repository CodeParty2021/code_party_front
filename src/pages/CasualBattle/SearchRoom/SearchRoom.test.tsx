import React from "react";
import { shallow } from "enzyme";

import { CasualBattleSearchRoom } from "./SearchRoom";
import { IResponse, useSearchRoomState } from "./hooks/useSearchRoomState";

jest.mock("./hooks/useSearchRoomState");

const useSelectorMock = useSearchRoomState as jest.Mock;

const state: IResponse = {
  enterBtnDisabled: false,
  enterBtnClickHandler: jest.fn(),
  roomIdTextBoxChangeHandler: jest.fn(),
  roomIdTextBoxValue: "typed value",
};

describe("<CasualBattleSearchRoom />", () => {
  beforeEach(() => {
    useSelectorMock.mockReturnValue({
      ...state,
    });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("snapshot test", () => {
    const wrapper = shallow(<CasualBattleSearchRoom />);

    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it("enter button test", () => {
    const wrapper = shallow(<CasualBattleSearchRoom />);

    const btn = wrapper.find("#enter-btn");
    btn.simulate("click");

    expect(state.enterBtnClickHandler).toHaveBeenCalled();
  });

  it("room id text box test", () => {
    const wrapper = shallow(<CasualBattleSearchRoom />);

    const event = {
      target: {
        value: "update value",
      },
    };

    const btn = wrapper.find("#roomid-textbox");
    btn.simulate("change", event);

    expect(state.roomIdTextBoxChangeHandler).toHaveBeenCalled();
  });
});
