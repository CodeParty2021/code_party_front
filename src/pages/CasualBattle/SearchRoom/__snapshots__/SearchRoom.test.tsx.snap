// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CasualBattleSearchRoom /> snapshot test 1`] = `
Array [
  <div>
    <div>
      ルームIDを入力する
    </div>
    <input
      id="roomid-textbox"
      onChange={[Function]}
      value="typed value"
    />
    <button
      disabled={false}
      id="enter-btn"
      onClick={[MockFunction]}
    >
      ルームに入る
    </button>
  </div>,
]
`;
